src/System.Private.ServiceModel/src/System/ServiceModel/Channels/CommunicationObject.cs

```diff
@@ -769,15 +773,23 @@ internal protected virtual Task OnCloseAsync(TimeSpan timeout)
 #endif
 
             // If we have already executed this code path, just return a completed Task.
-            // It means this type did not override OnCloseAsync or OnBeginClose.
+            // This is a safeguard in case custom CommunicationObjects attempt to do this multiple times.
             if (_onCloseAsyncCalled)
             {
                 return TaskHelpers.CompletedTask();
             }
 
-            // External types cannot override this method because it is not public.
-            // Redirect this call to the APM path.
             _onCloseAsyncCalled = true;
+
+            // External types cannot override this method because it is not public.
+            // If the caller started this open synchronously, we must use the synchronous
+            // methods they may have overridden.
+            if (_isSynchronousClose)
+            {
+                return Task.Run(() => OnClose(timeout));
```

Comment:
Interesting - what's a good solution to this? 

---
In reply to: [79509865](https://github.com/dotnet/wcf/pull/1547#discussion_r79509865) [](ancestors = 79509865)